File: go/examples/real_world.go
//go:build ignore
// +build ignore

package main

import (
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"time"

	tracing "github.com/LeDuyViet/quicktrace/go"
)

// Simulate different types of operations with realistic timing
func simulateDatabase(tracer *tracing.Tracer, query string) map[string]interface{} {
	tracer.Span("Database: " + query)

	// Simulate database query time (realistic variability)
	baseTime := 50 + rand.Intn(100) // 50-150ms
	time.Sleep(time.Duration(baseTime) * time.Millisecond)

	return map[string]interface{}{
		"query":   query,
		"results": rand.Intn(100),
		"time":    baseTime,
	}
}

func simulateCache(tracer *tracing.Tracer, key string) bool {
	tracer.Span("Cache lookup: " + key)

	// Cache operations are usually very fast
	time.Sleep(time.Duration(1+rand.Intn(5)) * time.Millisecond)

	// 70% cache hit rate
	return rand.Float32() < 0.7
}

func simulateExternalAPI(tracer *tracing.Tracer, service string) map[string]interface{} {
	tracer.Span("External API: " + service)

	// External APIs can be slow and variable
	baseTime := 200 + rand.Intn(800) // 200ms-1s
	time.Sleep(time.Duration(baseTime) * time.Millisecond)

	return map[string]interface{}{
		"service": service,
		"status":  "success",
		"latency": baseTime,
	}
}

func simulateBusinessLogic(tracer *tracing.Tracer, complexity string) {
	tracer.Span("Business logic: " + complexity)

	var duration int
	switch complexity {
	case "simple":
		duration = 5 + rand.Intn(15) // 5-20ms
	case "medium":
		duration = 20 + rand.Intn(50) // 20-70ms
	case "complex":
		duration = 100 + rand.Intn(200) // 100-300ms
	}

	time.Sleep(time.Duration(duration) * time.Millisecond)
}

// HTTP Handlers with tracing
func userProfileHandler(w http.ResponseWriter, r *http.Request) {
	tracer := tracing.NewSimpleTracer("GET /api/user/profile",
		tracing.WithOutputStyle(tracing.StyleDetailed),
		tracing.WithHideUltraFast(2*time.Millisecond))

	userID := r.URL.Query().Get("id")
	if userID == "" {
		userID = "12345"
	}

	// Check cache first
	cacheHit := simulateCache(tracer, "user:"+userID)

	var userData map[string]interface{}
	if !cacheHit {
		// Cache miss - query database
		userData = simulateDatabase(tracer, "SELECT * FROM users WHERE id = "+userID)

		// Cache the result
		tracer.Span("Cache store")
		time.Sleep(2 * time.Millisecond)
	} else {
		tracer.Span("Cache hit")
		time.Sleep(1 * time.Millisecond)
		userData = map[string]interface{}{
			"id":     userID,
			"name":   "John Doe",
			"cached": true,
		}
	}

	// Process business logic
	simulateBusinessLogic(tracer, "medium")

	// Enrich with additional data
	enrichData := simulateExternalAPI(tracer, "user-enrichment-service")

	// Final response preparation
	tracer.Span("Response serialization")
	time.Sleep(3 * time.Millisecond)

	response := map[string]interface{}{
		"user":       userData,
		"enrichment": enrichData,
		"timestamp":  time.Now().Unix(),
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)

	tracer.End()
}

func analyticsHandler(w http.ResponseWriter, r *http.Request) {
	tracer := tracing.NewSimpleTracer("POST /api/analytics",
		tracing.WithOutputStyle(tracing.StyleColorful),
		tracing.WithShowSlowOnly(50*time.Millisecond),
		tracing.WithGroupSimilar(20*time.Millisecond))

	// Validate request
	tracer.Span("Request validation")
	time.Sleep(5 * time.Millisecond)

	// Multiple database queries for analytics
	for i := 1; i <= 3; i++ {
		query := fmt.Sprintf("analytics_query_%d", i)
		simulateDatabase(tracer, query)
	}

	// Complex data processing
	simulateBusinessLogic(tracer, "complex")

	// Generate multiple reports
	for i := 1; i <= 4; i++ {
		tracer.Span(fmt.Sprintf("Generate report %d", i))
		time.Sleep(time.Duration(30+rand.Intn(40)) * time.Millisecond)
	}

	// Store results
	simulateDatabase(tracer, "INSERT INTO analytics_results")

	response := map[string]interface{}{
		"status":            "completed",
		"reports_generated": 4,
		"timestamp":         time.Now().Unix(),
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)

	tracer.End()
}

func healthCheckHandler(w http.ResponseWriter, r *http.Request) {
	tracer := tracing.NewSimpleTracer("GET /health",
		tracing.WithOutputStyle(tracing.StyleMinimal),
		tracing.WithMinTotalDuration(10*time.Millisecond)) // Only show if > 10ms

	// Quick database ping
	tracer.Span("Database ping")
	time.Sleep(time.Duration(2+rand.Intn(8)) * time.Millisecond)

	// Quick cache ping
	tracer.Span("Cache ping")
	time.Sleep(time.Duration(1+rand.Intn(3)) * time.Millisecond)

	// External service check
	tracer.Span("External service check")
	time.Sleep(time.Duration(5+rand.Intn(15)) * time.Millisecond)

	response := map[string]interface{}{
		"status":    "healthy",
		"timestamp": time.Now().Unix(),
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(response)

	tracer.End()
}

func main() {
	fmt.Println("🌐 Real-World HTTP API Server with QuickTrace")
	fmt.Println("=============================================")

	// Seed random for consistent demo
	rand.Seed(time.Now().UnixNano())

	// Setup routes
	http.HandleFunc("/api/user/profile", userProfileHandler)
	http.HandleFunc("/api/analytics", analyticsHandler)
	http.HandleFunc("/health", healthCheckHandler)

	// Demo mode - simulate requests instead of starting server
	fmt.Println("\n🎬 Demo Mode - Simulating API Requests:")
	fmt.Println("======================================")

	// Simulate various API calls
	fmt.Println("\n1️⃣ User Profile Request (Medium complexity)")
	req1, _ := http.NewRequest("GET", "/api/user/profile?id=12345", nil)
	userProfileHandler(nil, req1)

	time.Sleep(1 * time.Second)

	fmt.Println("\n2️⃣ Analytics Request (High complexity)")
	req2, _ := http.NewRequest("POST", "/api/analytics", nil)
	analyticsHandler(nil, req2)

	time.Sleep(1 * time.Second)

	fmt.Println("\n3️⃣ Health Check (Low complexity)")
	req3, _ := http.NewRequest("GET", "/health", nil)
	healthCheckHandler(nil, req3)

	time.Sleep(1 * time.Second)

	fmt.Println("\n4️⃣ Multiple Quick Health Checks")
	for i := 1; i <= 3; i++ {
		fmt.Printf("\nHealth check #%d:\n", i)
		req, _ := http.NewRequest("GET", "/health", nil)
		healthCheckHandler(nil, req)
		time.Sleep(500 * time.Millisecond)
	}

	fmt.Println("\n✅ Demo completed!")
	fmt.Println("\n💡 In a real server, uncomment the lines below:")
	fmt.Println("   // fmt.Println(\"Server starting on :8080\")")
	fmt.Println("   // log.Fatal(http.ListenAndServe(\":8080\", nil))")

	// Uncomment to run as real server
	// fmt.Println("🚀 Server starting on :8080")
	// log.Fatal(http.ListenAndServe(":8080", nil))
}

File: js/examples/real_world.js
const { Tracer } = require('../tracer.js');

// Mock Express.js like functionality for demo
class MockApp {
    constructor() {
        this.routes = {};
    }

    get(path, handler) {
        this.routes[`GET ${path}`] = handler;
    }

    post(path, handler) {
        this.routes[`POST ${path}`] = handler;
    }

    async simulate(method, path, query = {}) {
        const key = `${method} ${path}`;
        const handler = this.routes[key];
        if (handler) {
            const req = { method, path, query };
            const res = { 
                json: (data) => console.log('Response:', JSON.stringify(data, null, 2)),
                status: (code) => ({ json: (data) => console.log(`Status ${code}:`, JSON.stringify(data, null, 2)) })
            };
            await handler(req, res);
        }
    }
}

// Simulate different types of operations with realistic timing
async function simulateDatabase(tracer, query) {
    tracer.span(`Database: ${query}`);
    
    // Simulate database query time (realistic variability)
    const baseTime = 50 + Math.random() * 100; // 50-150ms
    await new Promise(resolve => setTimeout(resolve, baseTime));
    
    return {
        query: query,
        results: Math.floor(Math.random() * 100),
        time: baseTime
    };
}

async function simulateCache(tracer, key) {
    tracer.span(`Cache lookup: ${key}`);
    
    // Cache operations are usually very fast
    await new Promise(resolve => setTimeout(resolve, 1 + Math.random() * 5));
    
    // 70% cache hit rate
    return Math.random() < 0.7;
}

async function simulateExternalAPI(tracer, service) {
    tracer.span(`External API: ${service}`);
    
    // External APIs can be slow and variable
    const baseTime = 200 + Math.random() * 800; // 200ms-1s
    await new Promise(resolve => setTimeout(resolve, baseTime));
    
    return {
        service: service,
        status: 'success',
        latency: baseTime
    };
}

async function simulateBusinessLogic(tracer, complexity) {
    tracer.span(`Business logic: ${complexity}`);
    
    let duration;
    switch (complexity) {
        case 'simple':
            duration = 5 + Math.random() * 15; // 5-20ms
            break;
        case 'medium':
            duration = 20 + Math.random() * 50; // 20-70ms
            break;
        case 'complex':
            duration = 100 + Math.random() * 200; // 100-300ms
            break;
    }
    
    await new Promise(resolve => setTimeout(resolve, duration));
}

// HTTP Handlers with tracing
async function userProfileHandler(req, res) {
    const tracer = new Tracer('GET /api/user/profile', {
        style: 'detailed',
        hideUltraFast: 2 // ms
    });

    const userID = req.query.id || '12345';

    // Check cache first
    const cacheHit = await simulateCache(tracer, `user:${userID}`);
    
    let userData;
    if (!cacheHit) {
        // Cache miss - query database
        userData = await simulateDatabase(tracer, `SELECT * FROM users WHERE id = ${userID}`);
        
        // Cache the result
        tracer.span('Cache store');
        await new Promise(resolve => setTimeout(resolve, 2));
    } else {
        tracer.span('Cache hit');
        await new Promise(resolve => setTimeout(resolve, 1));
        userData = {
            id: userID,
            name: 'John Doe',
            cached: true
        };
    }

    // Process business logic
    await simulateBusinessLogic(tracer, 'medium');

    // Enrich with additional data
    const enrichData = await simulateExternalAPI(tracer, 'user-enrichment-service');
    
    // Final response preparation
    await new Promise(resolve => setTimeout(resolve, 3));
    tracer.span('Response serialization');

    const response = {
        user: userData,
        enrichment: enrichData,
        timestamp: Date.now()
    };

    res.json(response);
    tracer.end();
}

async function analyticsHandler(req, res) {
    const tracer = new Tracer('POST /api/analytics', {
        style: 'colorful',
        showSlowOnly: 50, // ms
        groupSimilar: 20 // ms
    });

    // Validate request
    await new Promise(resolve => setTimeout(resolve, 5));
    tracer.span('Request validation');

    // Multiple database queries for analytics
    for (let i = 1; i <= 3; i++) {
        const query = `analytics_query_${i}`;
        await simulateDatabase(tracer, query);
    }

    // Complex data processing
    await simulateBusinessLogic(tracer, 'complex');

    // Generate multiple reports
    for (let i = 1; i <= 4; i++) {
        await new Promise(resolve => setTimeout(resolve, 30 + Math.random() * 40));
        tracer.span(`Generate report ${i}`);
    }

    // Store results
    await simulateDatabase(tracer, 'INSERT INTO analytics_results');

    const response = {
        status: 'completed',
        reports_generated: 4,
        timestamp: Date.now()
    };

    res.json(response);
    tracer.end();
}

async function healthCheckHandler(req, res) {
    const tracer = new Tracer('GET /health', {
        style: 'minimal',
        minTotalDuration: 10 // ms - Only show if > 10ms
    });

    // Quick database ping
    await new Promise(resolve => setTimeout(resolve, 2 + Math.random() * 8));
    tracer.span('Database ping');

    // Quick cache ping
    await new Promise(resolve => setTimeout(resolve, 1 + Math.random() * 3));
    tracer.span('Cache ping');

    // External service check
    await new Promise(resolve => setTimeout(resolve, 5 + Math.random() * 15));
    tracer.span('External service check');

    const response = {
        status: 'healthy',
        timestamp: Date.now()
    };

    res.json(response);
    tracer.end();
}

async function main() {
    console.log('🌐 Real-World HTTP API Server with QuickTrace (JavaScript)');
    console.log('==========================================================');

    const app = new MockApp();

    // Setup routes
    app.get('/api/user/profile', userProfileHandler);
    app.post('/api/analytics', analyticsHandler);
    app.get('/health', healthCheckHandler);

    // Demo mode - simulate requests
    console.log('\n🎬 Demo Mode - Simulating API Requests:');
    console.log('======================================');

    // Simulate various API calls
    console.log('\n1️⃣ User Profile Request (Medium complexity)');
    await app.simulate('GET', '/api/user/profile', { id: '12345' });

    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('\n2️⃣ Analytics Request (High complexity)');
    await app.simulate('POST', '/api/analytics');

    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('\n3️⃣ Health Check (Low complexity)');
    await app.simulate('GET', '/health');

    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('\n4️⃣ Multiple Quick Health Checks');
    for (let i = 1; i <= 3; i++) {
        console.log(`\nHealth check #${i}:`);
        await app.simulate('GET', '/health');
        await new Promise(resolve => setTimeout(resolve, 500));
    }

    console.log('\n✅ Demo completed!');
    console.log('\n💡 In a real Express.js server, you would setup like:');
    console.log('   // const express = require(\'express\');');
    console.log('   // const app = express();');
    console.log('   // app.get(\'/api/user/profile\', userProfileHandler);');
    console.log('   // app.listen(3000, () => console.log(\'Server on port 3000\'));');
}

main().catch(console.error); 